### PR TITLE
Adding IAM support in Terraform for Dataform Repository resource

### PR DESCRIPTION
**Pull request by maverickjoy**
Adding IAM support in Terraform for Dataform Repository resource
Internal tracking bug: b/309433506

Product Resource API Documentation: https://cloud.google.com/dataform/reference/rest/v1beta1/projects.locations.repositories 

Did Manual Testing and it adds IAM policy to the repository added screenshots to the internal bug.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
dataform: added `google_dataform_repository_iam_*` IAM resources (beta)
```

### CODE REVIEW

**Review by SarahFrench — COMMENTED** [review]
Hi @maverickjoy , thanks for your contribution!

I can see that adding the YAML to trigger generation of the IAM resources has resulted in acceptance tests for those IAM resources being generated only in the Beta version of the provider. This is because [the first example](https://github.com/GoogleCloudPlatform/magic-modules/blob/6ba15516be5f190297781343222aa2576f33914e/mmv1/products/dataform/Repository.yaml#L34-L41) in this file is used to generate the IAM tests, and that example is marked as beta-only. Could you please check if that test is able to be promoted to the GA provider too?

There's currently a problem in the generated acceptance tests e.g. : 

```
iam_dataform_repository_generated_test.go:47:24 : fmt.Sprintf format %s reads arg #3, but call has 2 args
```

I believe the way to solve this is to add a `primary_resource_name` property to the first example, similar to [this example](https://github.com/GoogleCloudPlatform/magic-modules/blob/e80655649171a7fbe99ae7d251ec09bda6ab91f5/mmv1/products/bigquerydatapolicy/DataPolicy.yaml#L46-L47). The value of `primary_resource_name` will be templated into the generate tests and provide the missing argument to fmt.Sprintf

**Comment by maverickjoy** [comment]
> Hi @maverickjoy , thanks for your contribution!
> 
> I can see that adding the YAML to trigger generation of the IAM resources has resulted in acceptance tests for those IAM resources being generated only in the Beta version of the provider. This is because [the first example](https://github.com/GoogleCloudPlatform/magic-modules/blob/6ba15516be5f190297781343222aa2576f33914e/mmv1/products/dataform/Repository.yaml#L34-L41) in this file is used to generate the IAM tests, and that example is marked as beta-only. Could you please check if that test is able to be promoted to the GA provider too?
> 
> There's currently a problem in the generated acceptance tests e.g. :
> 
> ```
> iam_dataform_repository_generated_test.go:47:24 : fmt.Sprintf format %s reads arg #3, but call has 2 args
> ```
> 
> I believe the way to solve this is to add a `primary_resource_name` property to the first example, similar to [this example](https://github.com/GoogleCloudPlatform/magic-modules/blob/e80655649171a7fbe99ae7d251ec09bda6ab91f5/mmv1/products/bigquerydatapolicy/DataPolicy.yaml#L46-L47). The value of `primary_resource_name` will be templated into the generate tests and provide the missing argument to fmt.Sprintf

Thanks Sarah, for the review, honestly I am not sure if example should be added to GA since the resource is not published to GA itself ? as it min version is `beta` : https://github.com/GoogleCloudPlatform/magic-modules/blob/6ba15516be5f190297781343222aa2576f33914e/mmv1/products/dataform/Repository.yaml#L20

Thanks for the example, I have added it.

**Comment by maverickjoy** [comment]
> Also, I'm not a Googler so I won't be able to view the internal ticket I'm afraid! If there's any information relevant to this PR that can't be shared publicly let me know and we can figure out next steps

Nothing major as such, Its just some screenshots links which I had added in there, I can't attach them here, as the screenshot web page is internal to Google, so wont get opened from outside, but to summarise.
I generated the provider, ran `make build` on provider and then, created a terraform file and ran commands :

-   TF_CLI_CONFIG_FILE="./tf-dev-override.tfrc" terraform init -upgrade
-   TF_CLI_CONFIG_FILE="./tf-dev-override.tfrc" terraform plan
-    TF_CLI_CONFIG_FILE="./tf-dev-override.tfrc" terraform apply

And through curl command was able to see the IAM policy getting added for a user.

**Comment by SarahFrench** [comment]
> I am not sure if example should be added to GA since the resource is not published to GA itself 

Ah I didn't spot that the resource is only managed by a beta API, in that case yes lets leave the example as beta only and we can ignore the **Missing test report** automation.

**Review by SarahFrench — COMMENTED** [review]
Following my last comment, please add `min_version: beta` to the `iam_policy` block - this will mean the IAM resources are only generated in the Beta provider, similar to the Repository resource and the generated IAM acceptance tests being beta-only.

**Comment by maverickjoy** [comment]
> Following my last comment, please add `min_version: beta` to the `iam_policy` block - this will mean the IAM resources are only generated in the Beta provider, similar to the Repository resource and the generated IAM acceptance tests being beta-only.

For sure, Thank You

**Comment by SarahFrench** [comment]
Ok, there's a few problems happening with the acceptance tests.

Firstly, `TestAccDataformRepositoryIamMemberGenerated` & `TestAccDataformRepositoryIamPolicyGenerated` are failing because they both provision a secret with a static, hardcoded name, so tests will clash and cause errors like the one below: 

> Error: Error creating Secret: googleapi: Error 409: Secret [projects/[PROJECT_NUM]/secrets/secret] already exists.

This is due to an already existing problem with [the first example](https://github.com/GoogleCloudPlatform/magic-modules/blob/6ba15516be5f190297781343222aa2576f33914e/mmv1/products/dataform/Repository.yaml#L34-L41) in the Repository.yaml file. I've made a PR to fix this here https://github.com/GoogleCloudPlatform/magic-modules/pull/9462 that you could either pull into this PR or you could rebase your PR once it's merged to main.


Next, something is affecting  the `TestAccDataformRepositoryIamBindingGenerated` test's second step where it's tests performing `terraform import` on the google_dataform_repository_iam_binding resource. It's less clear what the problem is with this and I can do some manual testing to see if something is missing.

> vcr_utils.go:152: Failed state verification, resource with ID projects/[PROJECT_ID]/locations/us-central1/repositories/tf_test_repository4kcgewhaul/roles/viewer not found

**Comment by maverickjoy** [comment]
> Ok, there's a few problems happening with the acceptance tests.
> 
> 1. Firstly, `TestAccDataformRepositoryIamMemberGenerated` & `TestAccDataformRepositoryIamPolicyGenerated` are failing because they both provision a secret with a static, hardcoded name, so tests will clash and cause errors like the one below:
> 
> > Error: Error creating Secret: googleapi: Error 409: Secret [projects/[PROJECT_NUM]/secrets/secret] already exists.
> 
> This is due to an already existing problem with [the first example](https://github.com/GoogleCloudPlatform/magic-modules/blob/6ba15516be5f190297781343222aa2576f33914e/mmv1/products/dataform/Repository.yaml#L34-L41) in the Repository.yaml file. I've made a PR to fix this here #9462 that you could either pull into this PR or you could rebase your PR once it's merged to main.
> 
> 2. Next, something is affecting  the `TestAccDataformRepositoryIamBindingGenerated` test's second step where it tests performing `terraform import` on the google_dataform_repository_iam_binding resource. It's less clear what the problem is with this and I can do some manual testing to see if something is missing.
> 
> > vcr_utils.go:152: Failed state verification, resource with ID projects/[PROJECT_ID]/locations/us-central1/repositories/tf_test_repository4kcgewhaul/roles/viewer not found
> 
> Edit:
> 
> I did some manual tests about importing and found that I was able to import IAM resources fine, so I'm not sure why the `TestAccDataformRepositoryIamBindingGenerated` test is failing this way. I'll have another look tomorrow!

Sure, Thanks for looking into this, Ill also ping the guy for the PR you had raised, they may have some insights as well for this.

**Comment by SarahFrench** [comment]
That PR's just been merged into main, so if you rebase or pull main into your branch we can re-trigger the tests and see if anything related to the IAM resources requires fixing 🤞

File: mmv1/products/dataform/Repository.yaml [hunk 1]
```diff
@@ -25,12 +25,18 @@ references: !ruby/object:Api::Resource::ReferenceLinks
...
+  parent_resource_attribute: 'repository'
+  min_version: beta
 import_format:
   ['projects/{{project}}/locations/{{region}}/repositories/{{name}}']
 examples:
   - !ruby/object:Provider::Terraform::Examples
     name: 'dataform_repository'
     primary_resource_id: dataform_respository
+    primary_resource_name:
+      'fmt.Sprintf("tf_test_repository%s", context["random_suffix"])'
```
Comment (SarahFrench) [inline mmv1/products/dataform/Repository.yaml:39]: Looks like this is the cause of the confusion - the acceptance test creates a resource with the name `tf_test_dataform_repository<suffix>` but then tried to import IAM resources on a `tf_test_repository<suffix>` resource.

With the suggested change below I've had the tests pass locally: 

```suggestion
      'fmt.Sprintf("tf_test_dataform_repository%s", context["random_suffix"])'
```

Comment (SarahFrench) [reply mmv1/products/dataform/Repository.yaml]: I went ahead and made the commit so we can get the tests started - once they pass I'll approve and merge!

Comment (maverickjoy) [reply mmv1/products/dataform/Repository.yaml]: Thank You

**Review by SarahFrench — APPROVED** [review]
Tests pass! Merging now 🎉

**Pull request merged**